File: JobSearchCodeExamples/JobSearchCodeExamples.VB/TestTaker/ToInt.vb
Namespace TestTaker

    Public Class ToInt

        ''' <summary>
        ''' Romans to int.
        ''' </summary>
        ''' <paramname="s">The s.</param>
        ''' <returns></returns>
        Public Shared Function RomanToInt(ByVal s As String) As Integer

            Dim pos = 0
            Dim result = 0
            While pos < s.Length
                Dim spos As String = s(pos).ToString()
                Dim snext = String.Empty
                If pos < s.Length - 1 Then snext = s(pos + 1).ToString()
                Select Case spos
                    Case "I"
                        If Equals(snext, "V") Then
                            result += 4
                            pos += 1
                        ElseIf Equals(snext, "X") Then
                            result += 9
                            pos += 1
                        Else
                            result += 1
                        End If
                    Case "V"
                        result += 5
                    Case "X"
                        If Equals(snext, "L") Then
                            result += 40
                            pos += 1
                        ElseIf Equals(snext, "C") Then
                            result += 90
                            pos += 1
                        Else
                            result += 10
                        End If
                    Case "L"
                        result += 50
                    Case "C"
                        If Equals(snext, "D") Then
                            result += 400
                            pos += 1
                        ElseIf Equals(snext, "M") Then
                            result += 900
                            pos += 1
                        Else
                            result += 100
                        End If
                    Case "D"
                        result += 500
                    Case "M"
                        result += 1000

                End Select
                pos += 1
            End While
            Return result

        End Function

        ''' <summary>
        ''' Romans to int.
        ''' </summary>
        ''' <paramname="s">The s.</param>
        ''' <returns></returns>
        Public Shared Function RomanToInt2(ByVal s As String) As Integer

            Dim roman = {"I", "V", "X", "L", "C", "D", "M"}
            Dim values = {1, 5, 10, 50, 100, 500, 1000}
            Dim before = {"I", "X", "C"}
            Dim after = {New String() {"V", "X"}, New String() {"L", "C"}, New String() {"D", "M"}}
            Dim extraValues = {New Integer() {4, 9}, New Integer() {40, 90}, New Integer() {400, 900}}
            Dim pos = 0
            Dim result = 0
            While pos < s.Length
                Dim spos As String = s(pos).ToString()
                Dim snext = String.Empty
                If pos < s.Length - 1 Then snext = s(pos + 1).ToString()
                Dim romanIndex = Array.IndexOf(roman, spos)
                If before.Contains(spos) AndAlso Not String.IsNullOrEmpty(snext) Then
                    Dim beforeIndex = Array.IndexOf(before, spos)
                    If after(beforeIndex).Contains(snext) Then
                        result += extraValues(beforeIndex)(Array.IndexOf(after(beforeIndex), snext))
                        pos += 1
                    Else
                        result += values(romanIndex)
                    End If
                Else
                    result += values(romanIndex)
                End If
                pos += 1
            End While
            Return result

        End Function

        ''' <summary>
        ''' Romans to int.
        ''' </summary>
        ''' <paramname="s">The s.</param>
        ''' <returns></returns>
        Public Shared Function RomanToInt3(ByVal s As String) As Integer

            Dim RomanValues As New Dictionary(Of String, Int32) From {
            {"I", 1},
            {"V", 5},
            {"X", 10},
            {"L", 50},
            {"C", 100},
            {"D", 500},
            {"M", 1000}
        }
            Dim pos = 0
            Dim result = 0
            While pos < s.Length
                Dim spos As String = s(pos).ToString()
                Dim snext = String.Empty
                If pos < s.Length - 1 Then snext = s(pos + 1).ToString()
                If String.IsNullOrEmpty(snext) Then
                    result += RomanValues(spos)
                ElseIf RomanValues(spos) < RomanValues(snext) Then
                    result -= RomanValues(spos)
                Else
                    result += RomanValues(spos)
                End If
                pos += 1
            End While
            Return result
        End Function

    End Class

End Namespace

File: JobSearchCodeExamples/JobSearchCodeExamples/TestTaker/ToInt.cs
namespace JobSearchCodeExamples.TestTaker;

/// <summary>
/// Roman Numerals to an integer
/// </summary>
public static class ToInt
{
    /// <summary>
    /// Romans to int.
    /// </summary>
    /// <param name="s">The s.</param>
    /// <returns></returns>
    public static int RomanToInt(string s)
    {
        int pos = 0;
        int result = 0;
        while (pos < s.Length)
        {
            string spos = s[pos].ToString();
            string snext = string.Empty;
            if (pos < s.Length - 1)
                snext = s[pos + 1].ToString();
            switch (spos)
            {
                case "I":
                    if (snext == "V")
                    {
                        result += 4;
                        pos++;
                    }
                    else if (snext == "X")
                    {
                        result += 9;
                        pos++;
                    }
                    else
                        result += 1;
                    break;
                case "V":
                    result += 5;
                    break;
                case "X":
                    if (snext == "L")
                    {
                        result += 40;
                        pos++;
                    }
                    else if (snext == "C")
                    {
                        result += 90;
                        pos++;
                    }
                    else
                        result += 10;
                    break;
                case "L":
                    result += 50;
                    break;
                case "C":
                    if (snext == "D")
                    {
                        result += 400;
                        pos++;
                    }
                    else if (snext == "M")
                    {
                        result += 900;
                        pos++;
                    }
                    else
                        result += 100;
                    break;
                case "D":
                    result += 500;
                    break;
                case "M":
                    result += 1000;
                    break;

            }
            pos++;
        }
        return result;
    }

    /// <summary>
    /// Romans to int.
    /// </summary>
    /// <param name="s">The s.</param>
    /// <returns></returns>
    public static int RomanToInt2(string s)
    {
        string[] roman = { "I", "V", "X", "L", "C", "D", "M" };
        int[] values = { 1, 5, 10, 50, 100, 500, 1000 };
        string[] before = { "I", "X", "C" };
        string[][] after = { new string[] { "V", "X" }, new string[] { "L", "C" }, new string[] { "D", "M" } };
        int[][] extraValues = { new int[] { 4, 9 }, new int[] { 40, 90 }, new int[] { 400, 900 } };
        int pos = 0;
        int result = 0;
        while (pos < s.Length)
        {
            string spos = s[pos].ToString();
            string snext = string.Empty;
            if (pos < s.Length - 1)
                snext = s[pos + 1].ToString();
            int romanIndex = Array.IndexOf(roman, spos);
            if (before.Contains(spos) && !string.IsNullOrEmpty(snext))
            {
                int beforeIndex = Array.IndexOf(before, spos);
                if (after[beforeIndex].Contains(snext))
                {
                    result += extraValues[beforeIndex][Array.IndexOf(after[beforeIndex], snext)];
                    pos++;
                }
                else
                    result += values[romanIndex];
            }
            else
                result += values[romanIndex];
            pos++;
        }
        return result;
    }

    /// <summary>
    /// Romans to int.
    /// </summary>
    /// <param name="s">The s.</param>
    /// <returns></returns>
    public static int RomanToInt3(string s)
    {
        Dictionary<string, int> RomanValues = new();
        RomanValues.Add("I", 1);
        RomanValues.Add("V", 5);
        RomanValues.Add("X", 10);
        RomanValues.Add("L", 50);
        RomanValues.Add("C", 100);
        RomanValues.Add("D", 500);
        RomanValues.Add("M", 1000);
        int pos = 0;
        int result = 0;
        while (pos < s.Length)
        {
            string spos = s[pos].ToString();
            string snext = string.Empty;
            if (pos < s.Length - 1)
                snext = s[pos + 1].ToString();
            if (string.IsNullOrEmpty(snext))
                result += RomanValues[spos];
            else if (RomanValues[spos] < RomanValues[snext])
                result -= RomanValues[spos];
            else
                result += RomanValues[spos];
            pos++;
        }
        return result;
    }
}
